tmt/steps/prepare/__init__.py

```diff
@@ -221,6 +221,61 @@ def summary(self) -> None:
         preparations = fmf.utils.listed(self.preparations_applied, 'preparation')
         self.info('summary', f'{preparations} applied', 'green', shift=1)
 
+    def _extract_failed_packages(self, exceptions: list[Exception]) -> set[str]:
+        """
+        Extract package names from installation exceptions.
+
+        Returns a list of package names that failed to install.
+        """
+
+        failed_packages: set[str] = set()
+        import re
+
+        for exc in exceptions:
+            # Extract failed packages from RunError stderr/stdout
+            if isinstance(exc, tmt.utils.RunError):
+                error_text = ''
+                if exc.stderr:
+                    error_text += exc.stderr + '\n'
+                if exc.stdout:
+                    error_text += exc.stdout
+
+                if error_text:
+                    # dnf, dnf5, yum, apt, apk specific patterns
+                    patterns = [
+                        r'no package provides\s+([^\s\n]+)',
+                        r'Unable to find a match:\s+([^\s\n]+)',
+                        r'No match for argument:\s+([^\s\n]+)',
+                        r'No package\s+([^\s\n]+)\s+available',
+                        r'Unable to locate package\s+([^\s]+)',
+                        r'E:\s+Unable to locate package\s+([^\s]+)',
+                    ]
```

Comment:
Would it make sense to make these patterns package manager-specific? Let's say I'd add a plugin to support Gentoo package manager, `emerge`: I would need to update this list for `emerge` messages being decoded. Instead, we can add something like `extract_package_name_from_ugly_package_manager_output()` method, we can reach guest from here, and we can ask it to give us packages that somehow failed by calling the method with command output. Which patterns, stdout or stderr, that would all be owned by the package manager implementation. Plus, we don't need to run `apk` patterns for `dnf` output, and vice versa.

Second point: those patterns should be compiled, see various `re.compile()` calls on module level.

Reply:
:+1: for having the package managers owning these.